sklearn/utils/testing.py

```diff
@@ -697,5 +705,27 @@ def check_skip_travis():
     if os.environ.get('TRAVIS') == "true":
         raise SkipTest("This test needs to be skipped on Travis")
 
+class TempMemMap(object):
+    def __init__(self, X):
+        self.temp_folder = tempfile.mkdtemp()
+        self.X = X
+        self.X_read_only = None
+
+    def __enter__(self):
+        fpath = op.join(self.temp_folder, 'data.pkl')
+        joblib.dump(self.X, fpath)
+        self.X_read_only = joblib.load(fpath, mmap_mode='r')
+        return self.X_read_only
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._clean()
+
+    def _clean(self):
+        del self.X_read_only
```

Comment:
As we have no control on the scope of the variable that collects in the result of `__enter__` we have no guarantee on when the file descriptor of the `_mmap` buffer of the nested np.memmap instances will be closed. This means that the only way to properly collect the temporary files under windows it to use the following `atexit` based idiom:

https://github.com/joblib/joblib/blob/master/joblib/pool.py#L552